File: 2-medium/01038-binary-search-tree-to-greater-sum-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int sum;

    void convert_to_greater_sum(TreeNode* curr) {
        if (curr == nullptr) {
            return;
        }

        convert_to_greater_sum(curr->right);
        sum += curr->val;
        curr->val = sum;
        convert_to_greater_sum(curr->left);
    }
public:
    TreeNode* bstToGst(TreeNode* root) {
        sum = 0;
        convert_to_greater_sum(root);
        return root;
    }
};

File: 2-medium/01038-binary-search-tree-to-greater-sum-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def convert_to_greater_sum(self, curr: Optional[TreeNode]):
        if curr == None:
            return

        self.convert_to_greater_sum(curr.right)
        self.sum_values += curr.val
        curr.val = self.sum_values
        self.convert_to_greater_sum(curr.left)

    def bstToGst(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.sum_values = 0
        self.convert_to_greater_sum(root)
        return root

